pyomo/contrib/appsi/base.py

```diff
@@ -250,14 +253,26 @@ def __init__(self, primals, duals, slacks, reduced_costs):
         self._reduced_costs = reduced_costs
 
     def get_primals(self, vars_to_load: Optional[Sequence[_GeneralVarData]] = None) -> Mapping[_GeneralVarData, float]:
+        if self._primals is None:
+            raise RuntimeError(
+                'Solution loader does not currently have a valid solution. Please '
+                'check the termination condition.'
+            )
```

Comment:
Style question: this chunk of code seems to need to repeated a lot.  Would it make sense to channel @carldlaird and put this and other "universal" actions in the base class, then have derived classes implement `_get_primals_impl()` that the base `get_primals()` calls?